src/seq_mv/csr_spmv_device.c

```diff
@@ -101,75 +70,100 @@ hypre_csr_v_k_shuffle(hypre_DeviceItem &item,
          grid_row_id = grid_group_id;
       }
 
-      if (grid_group_id < n && group_lane < 2)
+      if (grid_group_id < nrows && group_lane < 2)
       {
          p = read_only_load(&d_ia[grid_row_id + group_lane]);
       }
       q = __shfl_sync(HYPRE_WARP_FULL_MASK, p, 1, K);
       p = __shfl_sync(HYPRE_WARP_FULL_MASK, p, 0, K);
 
-      T sum = 0.0;
-#if VERSION == 1
+      T sum[NV] = {T(0)};
+#if HYPRE_SPMV_VERSION == 1
 #pragma unroll 1
-      for (p += group_lane; warp_any_sync(item, HYPRE_WARP_FULL_MASK, p < q); p += K * 2)
+      for (p += group_lane; p < q; p += K * 2)
       {
-         if (p < q)
+         HYPRE_SPMV_ADD_SUM(p)
+         if (p + K < q)
          {
-            SPMV_ADD_SUM(p)
-            if (p + K < q)
-            {
-               SPMV_ADD_SUM((p + K))
-            }
+            HYPRE_SPMV_ADD_SUM((p + K))
          }
       }
-#elif VERSION == 2
+#elif HYPRE_SPMV_VERSION == 2
 #pragma unroll 1
       for (p += group_lane; warp_any_sync(item, HYPRE_WARP_FULL_MASK, p < q); p += K)
       {
          if (p < q)
          {
-            SPMV_ADD_SUM(p)
+            HYPRE_SPMV_ADD_SUM(p)
          }
       }
 #else
 #pragma unroll 1
       for (p += group_lane;  p < q; p += K)
       {
-         SPMV_ADD_SUM(p)
+         HYPRE_SPMV_ADD_SUM(p)
       }
 #endif
+
       // parallel reduction
 #pragma unroll
       for (HYPRE_Int d = K / 2; d > 0; d >>= 1)
       {
-         sum += __shfl_down_sync(HYPRE_WARP_FULL_MASK, sum, d);
+#pragma unroll
+         for (HYPRE_Int i = 0; i < NV; i++)
```

Comment:
@victorapm I have the same Q. as Rob asked. Do you think this for-loop (NV=1) has no effect and thus no overhead?

Reply:
I haven't seen any overhead, I'll post the results here

Reply:
OK. Glad to see you have the timing results. I guess the compiler is smart enough to optimize `for (HYPRE_Int i = 0; i < NV; i++)`, when `NV==1` at compile time.

Reply:
Yes! I would say that `#pragma unroll` is doing this optimization

Reply:
> Yes! I would say that `#pragma unroll` is doing this optimization

Good point